core/src/main/java/org/elasticsearch/common/cache/Cache.java

```diff
@@ -0,0 +1,642 @@
+ *     and applying to the LRU list.</li>
+ *     <li>Promotions could be deferred for entries that were "recently" promoted.</li>
+ *     <li>Locks on the list could be taken per node being modified instead of globally.</li>
+ * </ol>
+ *
+ * Evictions only occur after a mutation to the cache (meaning an entry promotion, a cache insertion, or a manual
+ * invalidation) or an explicit call to {@link #refresh()}.
+ *
+ * @param <K> The type of the keys
+ * @param <V> The type of the values
+ */
+public class Cache<K, V> {
+    // positive if entries have an expiration
+    private long expireAfter = -1;
+
+    // the number of entries in the cache
+    private int count = 0;
+
+    // the weight of the entries in the cache
+    private long weight = 0;
+
+    // the maximum weight that this cache supports
+    private long maximumWeight = -1;
+
+    // the weigher of entries
+    private ToLongBiFunction<K, V> weigher = (k, v) -> 1;
+
+    // the removal callback
+    private RemovalListener<K, V> removalListener = notification -> {
+    };
+
+    // use CacheBuilder to construct
+    Cache() {
+    }
+
+    void setExpireAfter(long expireAfter) {
+        if (expireAfter <= 0) {
+            throw new IllegalArgumentException("expireAfter <= 0");
+        }
+        this.expireAfter = expireAfter;
+    }
+
+    void setMaximumWeight(long maximumWeight) {
+        if (maximumWeight < 0) {
+            throw new IllegalArgumentException("maximumWeight < 0");
+        }
+        this.maximumWeight = maximumWeight;
+    }
+
+    void setWeigher(ToLongBiFunction<K, V> weigher) {
+        Objects.requireNonNull(weigher);
+        this.weigher = weigher;
+    }
+
+    void setRemovalListener(RemovalListener<K, V> removalListener) {
+        this.removalListener = removalListener;
+    }
+
+    /**
+     * The relative time used to track time-based evictions.
+     *
+     * @return the current relative time
+     */
+    protected long now() {
+        // System.nanoTime takes non-negligible time, so we only use it if we need it
+        return expireAfter == -1 ? 0 : System.nanoTime();
+    }
+
+    // the state of an entry in the LRU list
+    enum State {NEW, EXISTING, DELETED}
+
+    static class Entry<K, V> {
+        final K key;
+        final V value;
+        long accessTime;
+        Entry<K, V> before;
+        Entry<K, V> after;
+        State state = State.NEW;
+
+        public Entry(K key, V value, long accessTime) {
+            this.key = key;
+            this.value = value;
+            this.accessTime = accessTime;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null) {
+                return false;
+            } else if (!(obj instanceof Entry)) {
+                return false;
+            } else {
+                @SuppressWarnings("unchecked")
+                Entry<K, V> e = (Entry<K, V>) obj;
+                return Objects.equals(key, e.key);
+            }
```

Comment:
These are the result of an initial design which I subsequently removed. I'm pretty sure this is no longer needed at all. I've removed them in 8c05d4f43da9bfbba0173563ce22d9754f9b3dec.